### PR TITLE
support resizeable read/write memory

### LINKED ISSUE #1694: feat: support read/write memory initialisation
## Summary

Currently, read/write memory has to have every location initialised before it is read.  That is, any attempt to read an unitialised memory location results in a `panic`.  However, in principle, we could return `0` for uninitialised locations.

## Example

Here's a minimal test example:

```
pub input data(address:u16) -> (byte:u32)

fn main() {
    var x:u32 = data[0]
    var y:u32 = data[1]

    if x != y {
        fail
    }
}
```

With the following input:

```
{ "data": "0x" }
```

This causes a `panic`.

### PR DESCRIPTION
**Pull request by DavePearce**
This tweaks the way in which read/write memory handles out-of-bounds accesses (i.e. reads).  Previously, this would panic.  However, now it returns 0.

---

> [!NOTE]
> **Medium Risk**
> Changes core VM memory read semantics for RAM, which can affect program behavior and mask previously-failing invalid accesses; test coverage is added but reviewers should sanity-check edge cases around partial reads/padding.
> 
> **Overview**
> ZKC VM memory now distinguishes between `StaticArray` (fixed-size semantics; out-of-bounds reads still panic) and a new `DynamicArray` that **pads out-of-bounds reads with zeroes**.
> 
> `RandomAccess` (RAM) is switched to use `DynamicArray`, so reads past the currently-written region no longer crash and instead return zero (including via `FrameRead`). A new `basic_20` unit test program and golden output cover this behavior.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit ce76662672cd12d9280ad58a35ccd399dc05478f. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: pkg/zkc/vm/memory/dynamic_array.go [hunk 1]
```diff
@@ -0,0 +1,86 @@
...
+		zero   W
+		values []W
+	)
+	//
+	if start < n {
+		// Partially out-of-bounds
+		values = p.data[start:n]
+	}
+	// done
+	return array.BackPad(values, uint(start-end), zero)
```
[the remark this answers is not in the document]
Comment (DavePearce) [reply pkg/zkc/vm/memory/dynamic_array.go]: fixed
